File: backend/src/main/java/audioalchemy/engine/AIMusicEngine.java
package audioalchemy.engine;

import audioalchemy.model.AIMusicRequest;
import audioalchemy.model.AIMusicResponse;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class AIMusicEngine {

    public static AIMusicResponse analyze(AIMusicRequest request) {
        String combined = (
                safe(request.getPrompt()) + " " +
                        safe(request.getGenre()) + " " +
                        safe(request.getMood())
        ).toLowerCase();

        String mood = "neutral";
        String scale = "major";
        int tempo = request.getBpm() > 0 ? request.getBpm() : 100;
        String instrument = "piano";
        String chordProgression = "C - G - Am - F";
        List<String> melodyNotes = new ArrayList<>(Arrays.asList("C4", "E4", "G4", "A4", "G4", "E4", "C4"));

        if (combined.contains("sad") || combined.contains("melancholic") || combined.contains("emotional")) {
            mood = "sad";
            scale = "minor";
            tempo = Math.max(60, tempo - 20);
            instrument = "piano";
            chordProgression = "Am - F - C - G";
            melodyNotes = new ArrayList<>(Arrays.asList("A4", "C5", "E5", "D5", "C5", "A4", "G4"));
        }

        if (combined.contains("happy") || combined.contains("joyful") || combined.contains("bright")) {
            mood = "happy";
            scale = "major";
            tempo = Math.min(160, tempo + 15);
            instrument = "piano";
            chordProgression = "C - G - Am - F";
            melodyNotes = new ArrayList<>(Arrays.asList("C4", "D4", "E4", "G4", "A4", "G4", "E4"));
        }

        if (combined.contains("calm") || combined.contains("relaxing") || combined.contains("ambient")) {
            mood = "calm";
            scale = "major";
            tempo = Math.min(90, tempo);
            instrument = "piano";
            chordProgression = "C - F - Am - G";
            melodyNotes = new ArrayList<>(Arrays.asList("C4", "E4", "G4", "E4", "D4", "C4"));
        }

        if (combined.contains("heroic") || combined.contains("cinematic") || combined.contains("epic")) {
            mood = "heroic";
            scale = "major";
            tempo = Math.max(110, tempo);
            instrument = "orchestral piano";
            chordProgression = "C - Am - F - G";
            melodyNotes = new ArrayList<>(Arrays.asList("C4", "G4", "A4", "C5", "G4", "E4", "C4"));
        }

        if (combined.contains("eerie") || combined.contains("dark") || combined.contains("mysterious")) {
            mood = "eerie";
            scale = "minor";
            tempo = Math.min(85, tempo);
            instrument = "soft piano";
            chordProgression = "Dm - Bb - F - C";
            melodyNotes = new ArrayList<>(Arrays.asList("D4", "F4", "Ab4", "C5", "Ab4", "F4", "D4"));
        }

        if (combined.contains("electronic")) {
            instrument = "synth";
        } else if (combined.contains("fantasy")) {
            instrument = "bells";
        } else if (combined.contains("lo-fi")) {
            instrument = "electric piano";
        } else if (combined.contains("pop")) {
            instrument = "keyboard";
        }

        String title = buildTitle(request.getGenre(), mood);
        String description = buildDescription(request, mood, scale, tempo, instrument, chordProgression);

        return new AIMusicResponse(
                title,
                description,
                mood,
                scale,
                tempo,
                instrument,
                chordProgression,
                melodyNotes,
                "generated"
        );
    }

    private static String buildTitle(String genre, String mood) {
        String safeGenre = safe(genre).isBlank() ? "Custom" : capitalize(genre);
        String safeMood = mood.isBlank() ? "Track" : capitalize(mood);
        return "AI Track - " + safeGenre + " " + safeMood;
    }

    private static String buildDescription(AIMusicRequest request, String mood, String scale,
                                           int tempo, String instrument, String chordProgression) {
        return "Prompt: " + safe(request.getPrompt()) +
                " | Genre: " + safe(request.getGenre()) +
                " | Mood: " + mood +
                " | Scale: " + scale +
                " | Tempo: " + tempo +
                " BPM | Instrument: " + instrument +
                " | Chords: " + chordProgression +
                " | Duration: " + request.getDuration() + " sec" +
                " | " + (request.isInstrumentalOnly() ? "Instrumental only" : "Vocals allowed");
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    private static String capitalize(String value) {
        if (value == null || value.isBlank()) return "";
        return value.substring(0, 1).toUpperCase() + value.substring(1).toLowerCase();
    }
}

File: frontend/src/utils/aiMusicEngine.js
export function analyzeMusicRequest(input) {
    const prompt = (input.prompt || "").toLowerCase();
    const genre = input.genre || "cinematic";
    const moodInput = input.mood || "neutral";
    const bpmInput = Number(input.bpm) || 100;
    const duration = Number(input.duration) || 20;
    const instrumentalOnly = !!input.instrumentalOnly;

    let mood = moodInput;
    let scale = "major";
    let tempo = bpmInput;
    let instrument = "piano";
    let chordProgression = "C - G - Am - F";
    let melodyNotes = ["C4", "E4", "G4", "A4"];

    if (prompt.includes("sad") || moodInput.toLowerCase().includes("sad")) {
        mood = "sad";
        scale = "minor";
        tempo = Math.max(60, bpmInput - 20);
        instrument = "piano";
        chordProgression = "Am - F - C - G";
        melodyNotes = ["A4", "C5", "E5", "G5"];
    }

    if (prompt.includes("happy") || moodInput.toLowerCase().includes("happy")) {
        mood = "happy";
        scale = "major";
        tempo = Math.max(110, bpmInput);
        instrument = "piano";
        chordProgression = "C - G - Am - F";
        melodyNotes = ["C5", "D5", "E5", "G5"];
    }

    if (prompt.includes("calm") || moodInput.toLowerCase().includes("calm")) {
        mood = "calm";
        scale = "major";
        tempo = Math.min(80, bpmInput || 80);
        instrument = "piano";
        chordProgression = "C - Am - F - G";
        melodyNotes = ["C4", "E4", "G4", "B4"];
    }

    if (prompt.includes("dark") || prompt.includes("tense")) {
        mood = "dark";
        scale = "minor";
        tempo = Math.max(70, bpmInput);
        instrument = "synth";
        chordProgression = "Dm - Bb - F - C";
        melodyNotes = ["D4", "F4", "A4", "C5"];
    }

    if (genre.toLowerCase() === "lofi") {
        instrument = "electric piano";
        chordProgression = "Dm7 - G7 - Cmaj7 - Am7";
        melodyNotes = ["D4", "F4", "A4", "C5"];
    } else if (genre.toLowerCase() === "cinematic") {
        instrument = "strings";
        chordProgression = "Am - F - C - G";
        melodyNotes = ["A3", "C4", "E4", "G4"];
    } else if (genre.toLowerCase() === "electronic") {
        instrument = "synth";
        chordProgression = "Fm - Db - Ab - Eb";
        melodyNotes = ["F4", "Ab4", "C5", "Eb5"];
    }

    const title = `Generated Track - ${capitalize(genre)} ${capitalize(mood)}`;
    const description =
        `Prompt: ${input.prompt || ""} | Genre: ${genre} | Mood: ${mood} | BPM: ${tempo} | Duration: ${duration} sec | ` +
        (instrumentalOnly ? "Instrumental only" : "Vocals allowed");

    return {
        title,
        description,
        mood,
        scale,
        tempo,
        instrument,
        chordProgression,
        melodyNotes,
        status: "generated"
    };
}

function capitalize(value) {
    if (!value) return "Custom";
    return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
}